File: Python/1209. Remove All Adjacent Duplicates in String II.py
class Solution:
    def removeDuplicates(self, s: str, k: int) -> str:
        s = list(s)
        counts = list()
        j = 0
        for i in range(len(s)):
            s[j] = s[i]
            if j == 0 or s[j] != s[j - 1]:
                counts.append(1)
            else:
                incremented = counts.pop() + 1
                if incremented == k:
                    j = j - k
                else:
                    counts.append(incremented)
            j += 1
        return ''.join(s[0:j])
                    

File: 2024/1209. Remove All Adjacent Duplicates in String II.cpp
class Solution {
public:
    string removeDuplicates(string s, int k) {
        vector<pair<char, int>> counts = {{'#', 0}};
        for (char c : s) {
            if (counts.back().first != c) {
                counts.push_back({c, 1});
            } else {
                counts.back().second++;
                if (counts.back().second == k) {
                    counts.pop_back();
                }
            }
        }
        string res;
        for (auto& p : counts) {
            res += string(p.second, p.first);
        }
        return res;
    }
};